### PR TITLE
Paddle-2.0 API directory migration

### PR DESCRIPTION
**Pull request by pangyoki**
### PR types

Others

### PR changes

APIs

### Describe

Paddle 2.0 mainly promotes dygraph mode. The directory level of dygraph mode needs to be moved up one level, and the APIs of static mode are moved to the 'static' directory.

1. remove imperative directory
```
# dynamic and static mode switch API
paddle.enable_imperative  -> paddle.disable_static 
paddle.disable_imperative -> paddle.enable_static
paddle.in_imperative_mode -> paddle.in_dynamic_mode

# dynamic graph related API
paddle.imperative.to_variable -> paddle.to_variable
paddle.imperative.enable -> deleted
paddle.imperative.guard -> deleted
paddle.imperative.grad -> paddle.grad
paddle.imperative.no_grad -> paddle.no_grad
paddle.imperative.save -> paddle.save
paddle.imperative.load -> paddle.load
paddle.imperative.BackwardStrategy -> paddle.BackwardStrategy

# data_parallel related API
paddle.imperative.ParallelEnv -> paddle.ParallelEnv
paddle.imperative.prepare_context -> paddle.prepare_context
paddle.imperative.DataParalell -> paddle.DataParallel

# learning_rate_decay related API
paddle.imperative.NoamDecay -> paddle.NoamDecay
paddle.imperative.PiecewiseDecay -> paddle.PiecewiseDecay
paddle.imperative.NaturalExpDecay -> paddle.NaturalExpDecay
paddle.imperative.ExponentialDecay -> paddle.ExponentialDecay
paddle.imperative.InverseTimeDecay -> paddle.InverseTimeDecay
paddle.imperative.PolynomialDecay -> paddle.PolynomialDecay
paddle.imperative.CosineDecay -> paddle.CosineDecay

# dy_to_static related API
paddle.imperative.jit -> paddle.jit
paddle.imperative.TracedLayer -> paddle.jit.TracedLayer
paddle.imperative.declarative -> paddle.jit.to_static
paddle.imperative.ProgramTranslator -> paddle.jit.PragramTranslator
paddle.imperative.TranslatedLayer -> paddle.jit.TranslatedLayer
paddle.imperative.jit.save -> paddle.jit.save
paddle.imperative.jit.load -> paddle.jit.load
paddle.imperative.jit.SaveLoadConfig -> paddle.jit.SaveLoadConfig
```

2. Static graph related API in `paddle.framework`, moved to `paddle.static`
```
paddle.save -> paddle.static.save
paddle.load -> paddle.static.load

paddle.Executor -> paddle.static.Executor
paddle.global_scope -> paddle.static.global_scope
paddle.scope_guard -> paddle.static.scope_guard
paddle.append_backward -> paddle.static.append_backward
paddle.gradients -> paddle.static.gradients
paddle.BuildStrategy -> paddle.static.BuildStrategy
paddle.CompiledProgram -> paddle.static.CompiledProgram
paddle.ExecutionStrategy -> paddle.static.ExecutionStrategy
paddle.default_main_program -> paddle.static.default_main_program
paddle.default_startup_program -> paddle.static.default_startup_program
paddle.Program -> paddle.static.Program
paddle.name_scope -> paddle.static.name_scope
paddle.program_guard -> paddle.static.program_guard
paddle.Print -> paddle.static.Print
paddle.py_func -> paddle.static.py_func
paddle.ParallelExecutor -> paddle.static.ParallelExecutor
paddle.WeightNormParamAttr -> paddle.static.WeightNormParamAttr
```

3. Static graph related API in `paddle.declarative`, moved to `paddle.static.nn` 
```
paddle.declarative.fc -> paddle.static.nn.fc
paddle.declarative.batch_norm -> paddle.static.nn.batch_norm
paddle.declarative.bilinear_tensor_product -> paddle.static.nn.bilinear_tensor_product
paddle.declarative.conv2d -> paddle.static.nn.conv2d
paddle.declarative.conv2d_transpose -> paddle.static.nn.conv2d_transpose
paddle.declarative.conv3d -> paddle.static.nn.conv3d
paddle.declarative.conv3d_transpose -> paddle.static.nn.conv3d_transpose
paddle.declarative.create_parameter -> paddle.static.nn.create_parameter
paddle.declarative.crf_decoding -> paddle.static.nn.crf_decoding
paddle.declarative.data_norm -> paddle.static.nn.data_norm
paddle.declarative.deformable_conv -> paddle.static.nn.deformable_conv
paddle.declarative.group_norm -> paddle.static.nn.group_norm
paddle.declarative.hsigmoid -> paddle.static.nn.hsigmoid
paddle.declarative.instance_norm -> paddle.static.nn.instance_norm
paddle.declarative.layer_norm -> paddle.static.nn.layer_norm
paddle.declarative.multi_box_head -> paddle.static.nn.multi_box_head
paddle.declarative.nce -> paddle.static.nn.nce
paddle.declarative.prelu -> paddle.static.nn.prelu
paddle.declarative.row_conv -> paddle.static.nn.row_conv
paddle.declarative.spectral_norm -> paddle.static.nn.spectral_norm
paddle.declarative.embedding -> paddle.static.nn.embedding
```

### CODE REVIEW

File: python/paddle/__init__.py [hunk 1]
```diff
@@ -237,6 +244,10 @@
 
 from . import incubate
 from .incubate import hapi
-from .fluid.dygraph.base import enable_dygraph as enable_imperative  #DEFINE_ALIAS
-from .fluid.dygraph.base import disable_dygraph as disable_imperative  #DEFINE_ALIAS
-from .fluid.framework import in_dygraph_mode as in_imperative_mode  #DEFINE_ALIAS
+from .fluid.dygraph.base import enable_dygraph  #DEFINE_ALIAS
+from .fluid.dygraph.base import disable_dygraph  #DEFINE_ALIAS
+from .fluid.dygraph.base import enable_dygraph as disable_static  #DEFINE_ALIAS
+from .fluid.dygraph.base import disable_dygraph as enable_static  #DEFINE_ALIAS
```
Comment (zhiqiu) [inline python/paddle/__init__.py:244]: As discussed offline, `enable_dygraph` and `disable_static` both exist temporarily to pass the checks in `sampcd_processor.py`, which should be refined in next PR.

File: python/paddle/__init__.py [hunk 2]
```diff
@@ -201,30 +200,38 @@
...
 from .framework.random import manual_seed  #DEFINE_ALIAS
-from .framework import append_backward  #DEFINE_ALIAS
-from .framework import gradients  #DEFINE_ALIAS
-from .framework import Executor  #DEFINE_ALIAS
-from .framework import global_scope  #DEFINE_ALIAS
-from .framework import scope_guard  #DEFINE_ALIAS
-from .framework import BuildStrategy  #DEFINE_ALIAS
-from .framework import CompiledProgram  #DEFINE_ALIAS
 from .framework import default_main_program  #DEFINE_ALIAS
 from .framework import default_startup_program  #DEFINE_ALIAS
```
Comment (zhiqiu) [inline python/paddle/__init__.py]: Why `default_main_program` and `default_startup_program ` are kept?

File: python/paddle/fluid/install_check.py [hunk 3]
```diff
@@ -49,6 +50,10 @@ def run_check():
     This func should not be called only if you need to verify installation
     '''
     print("Running Verify Fluid Program ... ")
+    '''
+    paddle 2.0 directory migration
+    Need to modify dygraph method.
+    '''
```
Comment (zhiqiu) [inline python/paddle/fluid/install_check.py]: This comment can be removed?

File: python/paddle/__init__.py [hunk 4]
```diff
@@ -201,30 +200,38 @@
...
-from .framework import create_global_var  #DEFINE_ALIAS
-from .framework import create_parameter  #DEFINE_ALIAS
-from .framework import Print  #DEFINE_ALIAS
-from .framework import py_func  #DEFINE_ALIAS
-from .framework import ExecutionStrategy  #DEFINE_ALIAS
-from .framework import name_scope  #DEFINE_ALIAS
-from .framework import ParallelExecutor  #DEFINE_ALIAS
-from .framework import ParamAttr  #DEFINE_ALIAS
 from .framework import Program  #DEFINE_ALIAS
```
Comment (zhiqiu) [inline python/paddle/__init__.py]: same above

**Comment by zhiqiu** [comment]
Please add some unittests to check the migration.

File: python/paddle/fluid/dygraph/base.py [hunk 5]
```diff
@@ -121,8 +121,8 @@ def enabled():
 
 def enable_dygraph(place=None):
     """
-    :alias_main: paddle.enable_dygraph
-	:alias: paddle.enable_dygraph,paddle.enable_imperative.enable_dygraph
+    :alias_main: paddle.disable_static
+	:alias: paddle.disable_static, paddle.enable_dygraph
 	:old_api: paddle.fluid.dygraph.base.enable_dygraph
```
Comment (jzhang533) [inline python/paddle/fluid/dygraph/base.py]: alias_main, alias, old_api 标签可以删除。 下同。

**Review by jzhang533 — DISMISSED** [review]
lgtm
目前还是默认静态图模式，会再创建一个pr，改为默认的动态图模式。

File: python/paddle/fluid/dygraph/base.py (hunk 5 above)
Comment (pangyoki) [reply python/paddle/fluid/dygraph/base.py]: 已删除

File: python/paddle/__init__.py (hunk 2 above)
Comment (pangyoki) [reply python/paddle/__init__.py]: 已移至static目录下

File: python/paddle/fluid/install_check.py (hunk 3 above)
Comment (pangyoki) [reply python/paddle/fluid/install_check.py]: 已删除

File: python/paddle/fluid/tests/unittests/test_directory_migration.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            for module in old_directory:
+                run_cmd = self.get_import_command(module)
+                wb.write('try:\n')
+                wb.write('    ' + run_cmd + '\n')
+                wb.write('except:\n')
+                wb.write('    count += 1\n')
+                wb.write('else:\n')
+                wb.write('    err_module = "' + module + '"\n')
+            wb.write('if count != ' + str(len(old_directory)) + ':\n')
+            wb.write('    print("Error: Module " + err_module +' \
```
Comment (zhiqiu) [inline python/paddle/fluid/tests/unittests/test_directory_migration.py]: It's ok to write these lines, while I suggest you refine these lines by `string formating` for better readability in next PR. (maybe you already know that)
For example,

```
try_cmd_template = '''
try:
    {cmd}
except:
    count += 1
else:
    err_module={module}
'''

try_cmd = try_cmd_template.format(cmd="import paddle.xx", module="paddle.xx")
```

**Review by zhiqiu — COMMENTED** [review]
LGTM

**Review by luotao1 — DISMISSED** [review]

**Review by zhiqiu — DISMISSED** [review]
LGTM

**Review by lanxianghit — DISMISSED** [review]

**Review by XiaoguangHu01 — DISMISSED** [review]
LGTM

**Review by jzhang533 — DISMISSED** [review]
lgtm

File: python/paddle/fluid/tests/unittests/test_directory_migration.py (hunk 6 above)
Comment (pangyoki) [reply python/paddle/fluid/tests/unittests/test_directory_migration.py]: This format has been implemented in this PR.

**Review by XiaoguangHu01 — DISMISSED** [review]
LGTM

**Review by lanxianghit — DISMISSED** [review]

**Review by zhiqiu — DISMISSED** [review]

**Review by luotao1 — DISMISSED** [review]

**Review by jzhang533 — DISMISSED** [review]
lgtm

**Review by zhiqiu — DISMISSED** [review]

**Review by XiaoguangHu01 — DISMISSED** [review]
approved

**Review by jzhang533 — DISMISSED** [review]
lgtm

**Review by luotao1 — DISMISSED** [review]

File: python/paddle/__init__.py [hunk 7]
```diff
@@ -201,30 +200,34 @@
...
 from .framework import CUDAPlace  #DEFINE_ALIAS
 from .framework import CUDAPinnedPlace  #DEFINE_ALIAS
+
+from .framework import BackwardStrategy  #DEFINE_ALIAS
+from .framework import to_variable  #DEFINE_ALIAS
+from .framework import grad  #DEFINE_ALIAS
+from .framework import no_grad  #DEFINE_ALIAS
+from .framework import save_dygraph  #DEFINE_ALIAS
+from .framework import load_dygraph  #DEFINE_ALIAS
+from .framework import save  #DEFINE_ALIAS
```
Comment (raindrops2sea) [inline python/paddle/__init__.py:215]: conflict with from .tensor.io import save/load

**Review by raindrops2sea — CHANGES_REQUESTED** [review]

File: python/paddle/__init__.py (hunk 7 above)
Comment (pangyoki) [reply python/paddle/__init__.py:215]: save/load method of static graph mode has been moved to 'static/__init__.py'

**Review by zhiqiu — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]

**Review by jzhang533 — APPROVED** [review]
lgtm

**Review by luotao1 — APPROVED** [review]

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**